From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: derverdox <mail.ysp@web.de>
Date: Sun, 3 Dec 2023 14:20:30 +0100
Subject: [PATCH] Implementing Basic Item Format


diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index 4123f18aacae3c95841064404c1ddc75fbb8d1b7..9191b54aede3ce62da4c02e5d3d7c18c0609cd34 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -304,11 +304,13 @@ public final class ItemStack implements DataComponentHolder {
         this.count = count;
         this.components = components;
         this.getItem().verifyComponentsAfterLoad(this);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     private ItemStack(@Nullable Void v) {
         this.item = null;
         this.components = new PatchedDataComponentMap(DataComponentMap.EMPTY);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(getBukkitStack());
     }
 
     public static DataResult<Unit> validateComponents(DataComponentMap components) {
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 80e953be36bdc0c4a2988b797432010d486d08f2..13a03a5e07e0e0b9ca27ba59fe560db1e4a1eb08 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -169,6 +169,7 @@ public final class CraftItemStack extends ItemStack {
         this.setAmount(amount);
         this.setDurability(durability);
         this.setItemMeta(itemMeta);
+        de.verdox.itemformat.BasicItemFormat.applyItemFormatAndReturn(this);
     }
 
     @Override
diff --git a/src/test/java/org/bukkit/craftbukkit/inventory/DeprecatedItemMetaCustomValueTest.java b/src/test/java/org/bukkit/craftbukkit/inventory/DeprecatedItemMetaCustomValueTest.java
index 1c80fe7549d70ae16c7b755c22752549261f072a..3c8110dc83556aa1206ece71aedaaf910e939812 100644
--- a/src/test/java/org/bukkit/craftbukkit/inventory/DeprecatedItemMetaCustomValueTest.java
+++ b/src/test/java/org/bukkit/craftbukkit/inventory/DeprecatedItemMetaCustomValueTest.java
@@ -120,8 +120,8 @@ public class DeprecatedItemMetaCustomValueTest {
         String configValue = configuration.saveToString();
         YamlConfiguration loadedConfig = YamlConfiguration.loadConfiguration(new StringReader(configValue));
 
-        assertEquals(stack, loadedConfig.getSerializable("testpath", ItemStack.class));
-        assertNotEquals(new ItemStack(Material.DIAMOND), loadedConfig.getSerializable("testpath", ItemStack.class));
+        assertEquals(stack, de.verdox.itemformat.BasicItemFormat.removeConversionTag(loadedConfig.getSerializable("testpath", ItemStack.class)));
+        assertNotEquals(new ItemStack(Material.DIAMOND), de.verdox.itemformat.BasicItemFormat.removeConversionTag(loadedConfig.getSerializable("testpath", ItemStack.class)));
     }
 
     @Test
diff --git a/src/test/java/org/bukkit/craftbukkit/inventory/PersistentDataContainerTest.java b/src/test/java/org/bukkit/craftbukkit/inventory/PersistentDataContainerTest.java
index f33b49915d1f1f0838c49ac943e8d4d619450f6b..e9b3b61516c6a3c2f923dd0497d3cb8277c81ded 100644
--- a/src/test/java/org/bukkit/craftbukkit/inventory/PersistentDataContainerTest.java
+++ b/src/test/java/org/bukkit/craftbukkit/inventory/PersistentDataContainerTest.java
@@ -154,8 +154,8 @@ public class PersistentDataContainerTest {
         String configValue = configuration.saveToString();
         YamlConfiguration loadedConfig = YamlConfiguration.loadConfiguration(new StringReader(configValue));
 
-        assertEquals(stack, loadedConfig.getSerializable("testpath", ItemStack.class));
-        assertNotEquals(new ItemStack(Material.DIAMOND), loadedConfig.getSerializable("testpath", ItemStack.class));
+        assertEquals(stack, de.verdox.itemformat.BasicItemFormat.removeConversionTag(loadedConfig.getSerializable("testpath", ItemStack.class)));
+        assertNotEquals(new ItemStack(Material.DIAMOND), de.verdox.itemformat.BasicItemFormat.removeConversionTag(loadedConfig.getSerializable("testpath", ItemStack.class)));
     }
 
     @Test
